src/dotty/tools/dotc/reporting/diagnostic/messages.scala

```diff
@@ -678,4 +676,33 @@ object messages {
            |$fixedVarInAlternative""".stripMargin
     }
   }
+
+  case class IdentifierExpected()(implicit ctx: Context) extends Message(24) {
+    val kind = "Syntax"
+    val msg = "identifier expected"
+
+    val typeExplicitly = "def sum(): Int = {...}"
+
+    val typeImplicit = "def sum() = {...}"
+
+    val explanation = {
+      hl"""|When using type identifiers, you should pass a valid one.
+           |Because of Scala's stronger type, you can fix this issue either
+           |passing a valid type identifier or you can omit it, so the compiler can infer it for you.
+           |
+           |Let's write a function called ${"sum"}
+           |
+           |Here we explicitly pass a valid type identifier, in this case ${"Int"} for the function:
+           |
+           |$typeExplicitly
+           |
+           |But you also have the option to omit it completely.
+           |
+           |$typeImplicit
+           |
+           |This time the compiler will infer this type for you.
+           |
+           |""".stripMargin
+    }
+  }
```

Comment:
Ok, I'm going through your cases and explain to you what I thought right?

1 - Actually, to be honest, the only piece of code I could reproduce this error was this:

```
def foo: this = "bar"
```

and it goes to the `Parsers` class.

Is there any other case I could reproduce it, so I can make sure nothing was left behind?

2 - Get the user code, you mean I could use the `Tree` to show the user wrong code and give the correct one? In fact I could pass the `Tree` inside the class and use it. Don't know how to play with `Tree`s but I will get used to it.

3 - yeah, that one was difficult to write. I mean, Scala has a strong type system, which is good, everything has a type where it could be inferred by the compiler or not(in case the user provided it).

4 & 5 - No problem :)